test/e2e/all_in_one_test.go

```diff
@@ -298,6 +302,33 @@ func TestDeployAllInOnePostgresWithMultipleReplicas(t *testing.T) {
 	manifest, err := getTestManifest(t, postgresPath)
 	require.NoError(t, err)
 	deployment := deployKong(ctx, t, env, manifest)
+	// dump diagnostics and print out logs of KIC pod, if the test failed.
+	defer func() {
+		if t.Failed() {
+			output, err := cleaner.DumpDiagnostics(ctx, t.Name())
+			assert.NoError(t, err, "failed to dump diagnostics")
+			t.Logf("%s failed, dumped diagnostics to %s", t.Name(), output)
+			// print pod logs of KIC pods to debug when running in CI.
+			podLogDir := output + "/pod_logs"
+			logFiles, err := os.ReadDir(podLogDir)
+			assert.NoError(t, err, "failed to list files in pod log directory")
+			for _, logFile := range logFiles {
+				// print the log file if the pod belongs to KIC deployment
+				// by the pod name and namespace on the filename of log file.
+				if !strings.Contains(logFile.Name(), deployment.Namespace+"_"+deployment.Name) {
+					continue
+				}
+				if logFile.IsDir() {
+					continue
+				}
+
+				podLogContent, err := os.ReadFile(podLogDir + "/" + logFile.Name())
+				assert.NoErrorf(t, err, "failed to read logs in file %s", logFile.Name())
+				// replace the separator in filename `_` to `/`.
+				t.Logf("logs of pod %s:\n%s", strings.Replace(logFile.Name(), "_", "/", 1), string(podLogContent))
+			}
```

Comment:
dry: I'm OK with this for now, because we need something for better output, but I would like to see us in the future (maybe during our test reliability initiative upcoming in october?) taking some time to generalize some of this behavior so that we don't have to have so many long diagnostic blocks sprinkled throughout tests. Be thinking about ways we can be more DRY with this kind of code and create issues when ideas pop up.

Reply:
That is the intent of the change request in https://github.com/Kong/kubernetes-ingress-controller/pull/2747#issuecomment-1199839599, that we just handle it by invoking the cleaner, which is where all the actual diagnostic logic lives.

That's different in that it only gathers information once, at the time the test fails, but that's probably sufficient for most cases: Pod logs don't change if you collect them later, and cases where there's some conflicting config that happens to get removed by another test right as the cleaner runs are hopefully rare.

You could work continuous collection into the KTF system as well, I think: instead of the deferred cleaner diagnostics call, you could add a call at the beginning that spawns a goroutine on KTF's end, and have that goroutine run the diagnostic gather every N seconds. I don't know that we'd need that, and it'd probably something you only enable temporarily for very confusing flakes (it'd produce a ton of data to sift through, and produce data even when you don't need it), but it should be doable.

Reply:
It sounds like you're advocating that we don't need to move forward with this PR and that https://github.com/Kong/kubernetes-ingress-controller/pull/2747 should suffice?

Reply:
I glossed over the change and thought this was commentary on the original version of the PR still--this does incorporate the request from that comment, though normally you'd add the invocation to the same defer where you call cleanup, e.g.

```
defer func() {
	assert.NoError(t, env.Cleanup(ctx))
	if t.Failed() {
		output, err := cleaner.DumpDiagnostics(ctx, t.Name())
		assert.NoError(t, err, "failed to dump diagnostics")
		t.Logf("%s failed, dumped diagnostics to %s", t.Name(), output)
	}
}()
```
@randmonkey reading the logs from the temporary directory and echoing them isn't necessary; the action workflow will gather them and upload a zip to the test. E2E didn't have that enabled, but I've added it in https://github.com/Kong/kubernetes-ingress-controller/pull/2747/commits/7e0d317e5b2f2dfd69f82b875e9dea9b4bb91f4d.

Reply:
updated, removed the echoing of pod logs.